### DISCUSSION TITLE
Guidance around imlementing callbacks with custom deno embedding

### DISCUSSION THREAD

**Discussion by Yohe-Am** [denoland/deno#27504] [Q&A] [answered]
Hey, so we're embedding deno in our program with extensions added and have need for a callback system. I.e. the typescript code registers a bunch of functions that are supposed to be invoked by the native section. Calling a simple function is simple enough but some of the callbacks are to be async so it needs to support waiting on a promise.

We were able to find an example of how to get this done within the deno repositories, the [`watch_promise`](https://github.com/denoland/deno_core/blob/337223c154173a8505fb5fdbed836d33c58ef25d/core/runtime/bindings.rs#L911) construct from `deno_core`, and it generally works but with the exception of exception processing. The code adjacent to `watch_promise` relies on a lot of private machinery to handle this. I be I can hack around this with enough gumption but it seems like there should be an easier way to do this.

TLDR: what is the recommended way to implement callbacks or await js promises from rust side with support for catching exceptions?

**Comment by bartlomieju** [comment, marked as answer]
There's no built-in mechanism for this right now. When a promise rejects, it goes through machinery in `deno_core` that allows the unhandled rejection to be handled by callbacks like `onunhandledrejection` or `process.on("unhandledRejection")`.

**Reply by Yohe-Am** [reply]
Hey, thanks for the quick reply. I suppose I can use a javascript function to catch any unhandled errors and explicitly pipe that to the native side. 

But yeah, am I barking up the right tree with the `watch_promise` approach? I was not able to find any examples in the Deno extensions that relied on such a callback pattern. There seem to be a few methods on the `JsRuntime` type itself but they appear to coarse for this purpose.

**Reply by bartlomieju** [reply]
Yeah, `watch_promise` is the function you want to use and indeed you can use hooks in JS to catch unhandled promises to pipe it back to native side.
